rust/src/enip/detect.rs

```diff
@@ -0,0 +1,334 @@
+) -> bool {
+    let raw2: &CStr = CStr::from_ptr(raw); //unsafe
+    if let Ok(s) = raw2.to_str() {
+        if let Ok((_, v)) = enip_parse_command(s) {
+            *value = v;
+            return true;
+        }
+    }
+    return false;
+}
+
+fn enip_tx_is_cmd(
+    tx: &mut EnipTransaction, direction: Direction, value: u16,
+) -> std::os::raw::c_int {
+    if direction == Direction::ToServer {
+        if let Some(req) = &tx.request {
+            if req.header.cmd == value {
+                return 1;
+            }
+        }
+    } else {
+        if let Some(resp) = &tx.response {
+            if resp.header.cmd == value {
+                return 1;
+            }
+        }
+    }
+    return 0;
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn rs_enip_tx_is_cmd(
+    tx: *mut std::os::raw::c_void, direction: u8, value: u16,
+) -> std::os::raw::c_int {
+    let tx = cast_pointer!(tx, EnipTransaction);
+    return enip_tx_is_cmd(tx, direction.into(), value);
+}
+
+#[derive(Clone, Debug, Default)]
+pub struct DetectCipServiceData {
+    pub service: u8,
+    pub class: Option<u16>,
+    pub attribute: Option<u16>,
+}
+
+fn enip_parse_cip_service(i: &str) -> IResult<&str, DetectCipServiceData> {
+    let (i, _) = space0(i)?;
+    let (i, service) = verify(map_opt(digit1, |s: &str| s.parse::<u8>().ok()), |&v| {
+        v < 0x80
+    })(i)?;
+    let mut class = None;
+    let mut attribute = None;
+    let (i, _) = space0(i)?;
+    let (i, comma) = opt(char(','))(i)?;
+    let mut input = i;
+    if comma.is_some() {
+        let (i, _) = space0(i)?;
+        let (i, class1) = map_opt(digit1, |s: &str| s.parse::<u16>().ok())(i)?;
+        class = Some(class1);
+        let (i, _) = space0(i)?;
+        let (i, comma) = opt(char(','))(i)?;
+        input = i;
+        if comma.is_some() {
+            let (i, _) = space0(i)?;
+            let (i, negation) = opt(char('!'))(i)?;
+            let (i, attr1) = map_opt(digit1, |s: &str| s.parse::<u16>().ok())(i)?;
+            if negation.is_none() {
+                attribute = Some(attr1);
+            }
+            input = i;
+        }
+    }
+    let (i, _) = space0(input)?;
+    if !i.is_empty() {
+        return Err(nom7::Err::Error(make_error(i, ErrorKind::NonEmpty)));
+    }
+    return Ok((
+        i,
+        DetectCipServiceData {
+            service,
+            class,
+            attribute,
+        },
+    ));
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn rs_enip_parse_cip_service(
+    raw: *const std::os::raw::c_char,
+) -> *mut c_void {
+    let raw2: &CStr = CStr::from_ptr(raw); //unsafe
+    if let Ok(s) = raw2.to_str() {
+        if let Ok((_, ctx)) = enip_parse_cip_service(s) {
+            let boxed = Box::new(ctx);
+            return Box::into_raw(boxed) as *mut _;
+        }
+    }
+    return std::ptr::null_mut();
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn rs_enip_cip_service_free(ctx: *mut c_void) {
+    // Just unbox...
```

Comment:
Remove the comment, not accurate. But the code is good.

Reply:
ok